File: turtlesimMovement/turtlesimMotion.py
#!/usr/bin/env python
import rospy
from geometry_msgs.msg import Twist # to take linear and angular velocities
from turtlesim.msg import Pose # to take current position of robot in every sequence that we decided before.
import math # to calculate distance moved by the robot
import time # to make PID control 

# 4th step: functions that is needed for the movements have been written and coordinates for the Pose messages have been assigned.
x = 0
y = 0
yaw = 0 

def poseCallback (poseMessage): # This function is called every change in position happens.
    global x,y, yaw # to assert every changes in position we make these values up to date and readable in every parsion of the script
    x = poseMessage.x
    y = poseMessage.y
    yaw = poseMessage.theta

def moveRobot(speed,limitDistance,forward_backward):
    # 5th step: Twist object which carries velocity messages is called.
    velocityMessage = Twist()
    # 6th step: Current position must be called as global values
    global x,y
    x0 = x # initial x coordinate
    y0 = y # initial y coordinate
    # 7th step: Speed condition is decided according to moving forward or backward.
    if forward_backward == 1:
        velocityMessage.linear.x = abs(speed)
    else:
        velocityMessage.linear.x = -abs(speed)
    
    distanceMoved = 0.0 # before any movement distance value
    loopRate = rospy.Rate(10) # The messages are published 10 times a second.
    topicAdress = '/turtle1/cmd_vel'
    velocityPublisher = rospy.Publisher(topicAdress,Twist,queue_size = 10) # Publisher is created with topic which the publisher will use ,message type that is Twist
    # in this condition and queue size for the storing up to chosen (10) range into the buffer.
    while True: # Entering the loop that makes the actual moving 
        rospy.loginfo("Robot moves.") # The information is given to the terminal
        velocityPublisher.publish(velocityMessage) # Velocity messages are published
        loopRate.sleep() # This script is made for executing with proper loopRate()
        distanceMoved = abs(math.sqrt(pow((x-x0),2)+pow((-y0),2)))
        print(distanceMoved)
        if not (distanceMoved < limitDistance): # Condition that controls the exact final point is set.
            rospy.loginfo("Robot has reached the destination.")
            break
    
    # 9th step: Robot is stopped when the robot reached by giving values of velocity as 0 and finally this values must be published.
    velocityMessage.linear.x = 0.0
    velocityPublisher.publish(velocityMessage)

def rotateRobot(desiredAngularSpeedInDegree,desiredAngleInDegree,clckwise_cclckwsie):
    global yaw
    # Twist object is created and the values of these messages are set up to zero due to make initial values always in the same orientation.
    velocityMessage = Twist()
    velocityMessage.linear.x = 0.0
    velocityMessage.linear.y = 0.0
    velocityMessage.linear.z = 0.0
    velocityMessage.angular.x = 0.0
    velocityMessage.angular.y = 0.0
    velocityMessage.angular.z = 0.0

    angularSpeed = math.radians(desiredAngularSpeedInDegree) # this conversion is made due to difficulties of transforming degree to radian as a humanbeing.
    # This is an obligation because ROS uses radians.
    # 10th step: Angular speed is decided according to clockwise or counter-clockwise
    if clckwise_cclckwsie:
        velocityMessage.angular.z = -abs(angularSpeed)
    else:
        velocityMessage.angular.z = abs(angularSpeed)
    
    t0 = rospy.Time.now().to_sec() # to make time substraction first time is taken before the loop
    currentAngleInDegree = 0.0
    loopRate = rospy.Rate(15)
    topicAdress = '/turtle1/cmd_vel'
    velocityPublisher = rospy.Publisher(topicAdress,Twist,queue_size = 10)
    while True:
        rospy.loginfo("Robot is rotating.")
        velocityPublisher.publish(velocityMessage)
        t1 = rospy.Time.now().to_sec()
        currentAngleInDegree = (t1 - t0)*desiredAngularSpeedInDegree
        print(currentAngleInDegree)
        loopRate.sleep()
        if (currentAngleInDegree > desiredAngleInDegree):
            rospy.loginfo("Robot has reached desired angle.")
            break
    velocityMessage.angular.z = 0.0
    velocityPublisher.publish(velocityMessage)

def goToGoal(goalX,goalY):
    global x,y,yaw
    velocityMessage = Twist()   
    # PID controller was used in this method to achive less error when goal keeps closer.
    while True:
        kLinear = 0.5 # this coefficient can be changed according to system response
        distance = abs(math.sqrt(pow((goalX-x),2)+pow((goalY-y),2))) # In tis formula x and y are coming from Pose messages which are about coordinates 
        # of turtlesim
        linearSpeed = kLinear*distance # Proportional gain is used when the distance to the gal location decreaes ,speed also decreases.
        kAngular = 4.0
        angleOrientationToGoal = math.atan2(goalY-y,goalX-x) # Most proper angle is made for the goal 
        angularSpeed = (angleOrientationToGoal - yaw)*kAngular # When the angle exist with respcet to goal, angle decreases as the robot has a 
        # linear orientation o goal
        velocityMessage.linear.x = linearSpeed
        velocityMessage.angular.z = angularSpeed
        velocityPublisher.publish(velocityMessage)
        if distance < 0.01:
            break
                      
def setDesiredOrientation(desiredAngleInDegree):
    desiredAngleInRadians = math.radians(desiredAngleInDegree)
    exactAngleInRadians = desiredAngleInRadians - yaw
    if exactAngleInRadians < 0:
        clckwise_cclckwsie = 1
    else:
        clckwise_cclckwsie = 0
    print(exactAngleInRadians)
    print(desiredAngleInRadians)
    rotateRobot(20,math.degrees(exactAngleInRadians),clckwise_cclckwsie)

def gridMovement(): # in this movement almost every motion is tried to make similar real-world app.
    desiredPosition = Pose() # position info is taken to know coordinates that robot is standing and make the proper go to goal action
    #In position message , x,y and orientation which is theta are the desired infos for that method
    desiredPosition.x = 1
    desiredPosition.y = 1
    desiredPosition.theta = 0
    # All the values were set up by me, in these methods' values written in below are changable to test how outcomes affect
    goToGoal(8,8)
    setDesiredOrientation(math.degrees(desiredPosition.theta))
    moveRobot(2.0,3.0,True)
    rotateRobot(30,180,False)
    pass

def spiralMovement(): # In this method PID control was used.
    velocityMessage = Twist()
    currentPose = Pose()
    loopRate = rospy.Rate(10)
    topicAdress = '/turtle1/cmd_vel'
    velocityPublisher  rospy.Publisher(topicAdress,Twist,queue_size = 10)
    kAngularSpeed = 4
    kRadius = 0
    while((currentPose.x < 10.5) and (currentPose.y < 10.5)):
        # Current positions are limited for the turtlesim UI , It has almost x = 11 and y = 11 ranges in the coordinate system
        kRadius = kRadius + 1
        velocityMessage.linear.x = kRadius
        velocityMessage.linear.y = 0
        velocityMessage.linear.z = 0
        velocityMessage.angular.x = 0
        velocityMessage.angular.y = 0
        velocityMessage.angular.z = kAngularSpeed
        velocityPublisher.publish(velocityMessage)
        loopRate.sleep()
    velocityMessage.linear.x = 0
    velocityMessage.angular.x = 0
    velocityPublisher.publish(velocityMessage)

if __name__  == "__main__":
    try:
        # 1st step: Create a node
        nodeName = "cleanerRobotNode"
        rospy.init_node(nodeName,anonymous = True) # A node should have a unique name due to that reason anonymous = True
        # 2nd step: Subscriber and publisher are created
        velocityTopicAdress = '/turtle1/cmd_vel'
        velocityPublisher = rospy.Publisher(velocityTopicAdress,Twist,queue_size = 10)
        poseTopicAdress = '/turtle1/Pose'
        poseSubscriber = rospy.Subscriber(poseTopicAdress,Pose,poseCallback)
        time.sleep(2)
        # 3rd step: Methods which make movement are constructed
        # moveRobot(2.0,2.0,True)
        rotateRobot(25,45,True)
        # goToGoal(9,9)
        # setDesiredOrientation(180)
        # gridMovement()
        # spiralMovement()
    except rospy.ROSInterruptException:
        rospy.loginfo("Node has been terminated!")

File: turtlesimMovement/turtlesimMotion.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "turtlesim/Pose.h"
#include "sstream"

using namespace std;

global ros::Publisher velocityPublisher;
global ros::Subscriber poseSubscriber;
global turtlesim::Pose robotPosition;

const double xMin = 0.0;
const double yMin = 0.0;
const double xMax = 11.0;
const double yMax = 11.0;

const double PI = 3.14159265359;

void moveRobot(double speed, double distance, bool forward_backward);
void rotateRobot(double angularSpeedInRadians, double angleInDegreesInRadians, bool clckwise_cclckwise);
double degreesToRadians(double angleInDegrees);
void setDesiredOrientation(double angleInRadians);
void goToGoal(turtlesim::Pose goalPosition, double limitDistance);
void gridMovement();
void spiralMovement();
void poseCallback(const turtlesim::Pose::ConstPtr &poseMessage);

int main(int argc, char **argv)
{
    // ROS Node is created with giving a name, here is turtlesimMovement
    ros::init(argc, argv, "turtlesimMovement");
    ros::NodeHandle rosNodeHandle; // Noedhandle object is created to handle ros nodes
    double speed,angularSpeed;
    double distance,angleInDegreesInDegrees;
    bool forward_backward,clckwise_cclckwise;
    
    // Publisher and subscriber are created thanks to nodehandle
    velocityPublisher = rosNodeHandle.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel",1000);
    poseSubscriber = rosNodeHandle.subscribe("/turtle1/pose",10,poseCallback);

    ros::Rate loopRate(10);

    cout<<"Enter speed : ";
    cin>>speed;
    cout<<"Enter distance: ";
    cin>>distance;
    cout<<"Robot is moving forward/backward?";
    cin>>forward_backward;
    moveRobot(speed,distance,forward_backward);

    cout<<"Enter angular speed in degree/sec: ";
    cin>>angularSpeed;
    cout<<"Enter desired angleInDegreesInDegrees in degrees: ";
    cin>>angleInDegrees;
    cout<<"Robot is turning clockeise/counter-clockwise: ";
    cin>>clckwise_cclckwise;
    rotateRobot(degreesToRadians(angularSpeed),degreesToRadians(angleInDegrees),clckwise_cclckwise);

    setDesiredOrientation(degreesToRadians(180));

    // Pose object is created to adjust position of the robot fot the method goToGoal 
    turtlesim::Pose goalPosition;
    poseMesaage.x = 1;
    poseMessage.y = 1;
    poseMessag.theta = 0;
    goToGoal(goalPosition,0.01);    

    gridMovement();

    spiralMovement();

    loopRate.sleep();

    ros::spin();
    
    return 0;
    
    }

void moveRobot(double speed, double distance,bool forward_backward)
{
    geometry_msgs::Twist velocityMessage;
    if(forward_backward)
        velocityMessage.linear.x = abs(speed);
    else
        velocityMessage.linear.x = -abs(speed);

    velocityMessage.linear.y = 0;
    velocityMessage.linear.z = 0;
    velocityMessage.angular.x = 0;
    velocityMessage.angular.y = 0;
    velocityMesaage.angular.z = 0;
    double t0 = ros::Time::now().toSec();
    double distanceMoved = 0.0;
    ros::Rate loopRate(100);

    do{
        velocityPublisher.publish(velocityMessage);
        double t1 = ros::Time::now().toSec();
        distanceMoved = speed*(t1 - t10)
        ros::spinOnce();
        loopRate.sleep();
        cout<<(t1 - t0)<<","<<distanceMoved<<endl;      
    }while(distanceMoved < distance);
    velocityMessage.linear.x = 0;
    velocityPulisher.publish(velocityMessage);
}

void rotateRobot(double angularSpeedInRadians, double angleInDegreesInRadians, bool clckwise_cclckwise)
{
    geometry_msgs::Twist velocityMessage;
    velocityMessage.linear.x = 0;
    velocityMessage.linear.y = 0;
    velocityMessage.linear.z = 0;
    velocityMessage.angular.x = 0;
    velocityMessage.angular.y = 0;
    if(clckwise_cclckwise)
        velocityMessage.angular.z = -abs(angularSpeedInRadians);
    else
        velocityMessage.angular.z = abs(angularSpeedInRadians);

    double currentAngle = 0.0;
    double t0 = ros::Time::now().toSec();
    ros::Rate loopRate(10);
    do{
        velocityublisher.publish(velocityMessage);
        double t1 = ros::Time::now().toSec();
        currentAngle = angularSpeedInRadians*(t1 - t0);
        ros::spinOnce();
        loopRate.sleep();
    }while(currentAngle < angleInDegreesInRadians);
    velocityMessage.angular.z = 0;
    velocityPublisher.publish(velocityMessage);
}

double degreesToRadians(double angleInDegrees)
{
    return angleInDegrees*PI/180.0;
}

void setDesiredOrientation(double desiredAngleInRadians)
{
    double limitAngleInRadians = desiredAngleInRadians - robotPosition.theta;
    bool clckwise_cclckwise = ((limitAngleInRadians < 0)? true:false);
    cout<<desiredAngleInRadians<<","<<robotPosition.theta<<","<<limitAngleInRadians<<","<<clckwise_cclckwise<<endl;
    rotateRobot(degreesToRadians(10),abs(limitAngleInRadians),clckwise_cclckwise);
}

double getDistance(double x1, double y1, double x2,double y2)
{
    return sqrt(pow(x1 - x2),2) +pow((y1 -y2),2));
}

void goToGoal(turtlesim::Pose goalPosition, double limitDistance)
{
    geometry_msgs::Twist velocityMessage;
    ros::Rate loopRate(20);
    double distanceMoved = 0.0;
    do
    {
        double kLinear = 1.0;
        double kAngular = 4.0;
        double distance = getDistance(robotPosition.x,robotPosition.y,goalPosition.x,goalPosition.y);
        double distanceMoved = distanceMoved + distance;
        velocityMessage.linear.x = (kLinear*distance);
        velocityMessage.linear.y = 0;
        velocityMessage.linear.z = 0;
        velocityMessage.angular.x = 0;
        velocityMessage.angular.y = 0;
        desiredAngleToGoal = atan2(goalPosition.y - robotPosition.y,goalPosition.x - robotPosition.x)
        velocityMessage.angular.z = kAngular*(desiredAngleToGoal - robotPosition.theta);
        velocityPublisher.publish(velocityMessage);
        ros::spinOnce();
        loopRate.sleep();
    } while (getDistance(robotPosition.x,robotPosition.y,goalPosition.x,goalPosition.y) > limitDistance);
    cout<<"Robot has reached to the goal."<<endl;
    velocityMessage.linear.x = 0;
    velocityMessage.angular.z = 0;
    velocityPublisher.publish(velocityMessage); 
}

void gridMovement()
{
    ros::Rate loopRate(0.5);
    turtlesim::Pose goalPosition;
    goalPosition.x = 1;
    goalPosition.y = 1;
    goalPosition.theta = 0;
    goToGoal(goalPosition,0.01);
    loopRate.sleep();
    moveRobot(3.0,8.0,true);
    loopRate.sleep();
    rotateRobot(degreesToRadians(10),degreesToRadians(90),false);
    loopRate.sleep();
}

void spiralMovement()
{
    geometry_msgs::Twist velocityMessage;
    double spinCount = 0;
    double constantAngularSpeed = 4;
    double constantLinearSpeed = 0.5;
    ros::Rate loopRate(10);

    do
    { 
        constantLinearSpeed + = 1.0;
        velocityMessage.linear.x = constantLinearSpeed;
        velocityMessage.linear.y = 0;
        velocityMessage.linear.z = 0;
        velocityMessage.angular.x = 0;
        velocityMessage.angular.y = 0;
        velocityMessage.angular.z = constantAngularSpeed;
        cout<<"Linear velocity in x direction: "<<velocityMessage.linear.x<<endl;
        cout<<"Angular speed in z direction: "<<velocityMessage.angular.z<<endl;
        velocityPublisher.publish(velocityMessage);
        ros::spinOnce();
        loopRate.sleep();
    } while ((robotPosition.x < 10.0) && (rootPosition.y < 10.0));
    velocityMessage.linear.x = 0;
    velocityMessage.angular.z = 0;
    velocityPublisher.publish(velocityMessage);    
}

void poseCallback(const turtlesim::Pose::ConstPtr &poseMessage)
{
    robotPosition.x = poseMessage -> x;
    robotPosition.y = poseMessage -> y;
    robotPosition.theta = poseMessage -> theta;
}
